src/resilience/backend/FenixBackend.cpp

```diff
@@ -0,0 +1,299 @@
+    FENIX_SAFE_CALL( Fenix_Data_group_create( m_fenix_data_group_id, m_mpi_comm, 0, 0, m_fenix_policy_name, (void *)(m_fenix_policy_value), &m_fenix_policy_flag ) );
+  }
+
+  void
+  FenixMemoryBackend::register_hashes( const std::vector< KokkosResilience::ViewHolder > &views,
+                                       const std::vector< Detail::CrefImpl > &crefs  )
+  {
+    // Clear protected bits
+    for ( auto &&p : m_registry )
+    {
+      p.second.protect = false;
+    }
+
+    for ( auto &&view : views )
+    {
+      if ( !view->data() )  // uninitialized view
+        continue;
+
+      std::string label = get_canonical_label( view->label() );
+      auto iter = m_registry.find( label );
+
+      // Attempt to find the view in our registry
+      if ( iter == m_registry.end() )
+      {
+        // Calculate id using hash of view label
+        int id = ++m_last_id; // Prefix since we will consider id 0 to be no-id
+        iter = m_registry.emplace( std::piecewise_construct,
+                                        std::forward_as_tuple( label ),
+                                        std::forward_as_tuple( id ) ).first;
+        iter->second.element_size = view->data_type_size();
+        iter->second.size = view->span();
+
+        if ( !view->is_host_space() || !view->span_is_contiguous() )
+        {
+          // Can't reference memory directly, allocate memory for a watch buffer
+          iter->second.buff.assign( iter->second.size * iter->second.element_size, 0x00 );
+          iter->second.ptr = iter->second.buff.data();
+        } else {
+          iter->second.ptr = view->data();
+        }
+      }
+
+      // iter now pointing to our entry
+      iter->second.protect = true;
+    }
+
+    // Register crefs
+    for ( auto &&cref : crefs )
+    {
+      if ( !cref.ptr )  // uninitialized view
+        continue;
+      // If we haven't already register, register with VeloC
+      auto iter = m_registry.find( cref.name );
+      if ( iter == m_registry.end())
+      {
+        int id = ++m_last_id; // Prefix since we will consider id 0 to be no-id
+        iter = m_registry.emplace( std::piecewise_construct,
+            std::forward_as_tuple( cref.name ),
+            std::forward_as_tuple( id ) ).first;
+
+        iter->second.ptr = cref.ptr;
+        iter->second.size = cref.num;
+        iter->second.element_size = cref.sz;
+      }
+
+      iter->second.protect = true;
+    }
+
+    // Register everything protected, unregister anything unprotected
+    for ( auto &&p : m_registry )
+    {
+      if ( p.second.protect )
+      {
+        if ( !p.second.registered )
+        {
+          std::cout << "Protecting memory id " << p.second.id << " with label " << p.first << '\n';
+          FENIX_SAFE_CALL( Fenix_Data_member_create( m_fenix_data_group_id, p.second.id, p.second.ptr, p.second.size * p.second.element_size, MPI_CHAR ) );
+          p.second.registered = true;
+        }
```

Comment:
We should ensure the pointer is up-to-date
```suggestion
        } else {
          int flag;
          FENIX_SAFE_CALL( Fenix_Data_member_attr_set(
            m_fenix_data_group_id, p.second.id, FENIX_DATA_MEMBER_ATTRIBUTE_BUFFER, p.second.ptr, &flag
          );
          if(!flag) std::cerr << ...;
        }
```